### ISSUE TITLE
Bitstream does not run on 5CEFA5F23I7 , missing CMUXVG and CMUXHG statement ?

### ISSUE THREAD

**Issue by infphyny** [Ravenslofty/mistral#17]
In current state mistral is able to generate bitstream for 5CEFA5F23I7 / 5CEFA5F23C8. Unfortunately, the fpga is unable to run successfully the bitstream to sram and revert to previously loaded bitstream in flash produced by quartus.

When comparing decompiled bitstream between mistral and quartus there are missing CMUXVG and CMUXHG statement at the beginning of each bitstream for 5CEFA5F23I7 by mistral.

Also, when decompiling quartus generated bistream, mistral-cv output this

```
Source unknown on rnode SCLK.014.000.0015 ( 6, 01)
Source unknown on rnode SCLK.014.000.0020 ( 6, 04)
Source unknown on rnode SCLK.014.000.0025 ( 6, 08)
Source unknown on rnode SCLK.047.000.0005 ( 6, 08)
Source unknown on rnode SCLK.047.000.0009 ( 6, 02)
```
I have compared decompiled bistream between mistral and quartus for 3 examples [here](https://github.com/infphyny/nextpnr-mistral-issue)

In summary decompiled bistream generated by mistral begin like this
```
m 5CEFA5F23C8
o COMPRESSION_DIS 1
o OPT_B ffffff40.adffffff
r SCLK.047.000.0009 SCLKB1.047.001.0055
r SCLK.047.000.0009 SCLKB1.047.002.0054
r SCLK.047.000.0009 SCLKB1.047.003.0055
```
and decompiled bistream generated by quartus begin like this
```
m 5CEFA5F23C8
o COMPRESSION_DIS 1
o IDCODE 22
o JTAG_ID 0078fd4b
o OPT_B ffffff40.adffffff
r CMUXVG.028.000.0:CLKOUT SCLK.014.000.0000
r CMUXVG.028.061.0:CLKOUT SCLK.014.000.0001
r CMUXVG.028.000.3:CLKOUT SCLK.014.000.0005
r CMUXVG.028.061.1:CLKOUT SCLK.014.000.0010
r CMUXVG.028.000.1:CLKOUT SCLK.014.000.0012
r CMUXVG.028.061.0:CLKOUT SCLK.014.000.0013
r CMUXVG.028.061.3:CLKOUT SCLK.014.000.0014
r CMUXHG.000.008.0:CLKOUT SCLK.014.000.0022
r CMUXHG.000.008.3:CLKOUT SCLK.047.000.0007
r CMUXHG.000.008.0:CLKOUT SCLK.047.000.0008
r CMUXHG.068.008.3:CLKOUT SCLK.047.000.0010
r CMUXHG.000.008.3:CLKOUT SCLK.047.000.0020
r CMUXVG.028.000.1:CLKOUT SCLK.047.000.0021
r CMUXHG.068.008.2:CLKOUT SCLK.047.000.0024
r CMUXVG.028.000.2:CLKOUT SCLK.047.035.0027
r SCLK.047.035.0027 SCLKB1.047.047.0163
r SCLK.047.035.0027 SCLKB1.047.048.0163
```

I don't know if I can re-inject those CMUXHG CMUXVG from quartus to mistral, then ajusting numbers like 047.035.0027 for SCLK. Then recompile to rbf and try if it work on fpga.

Other than that, bitstream generated by mistral is very close to quartus.

Let me know if there are experiments I can do to try make mistral working on Cyclone V 5CEFA5F23 device.
Thanks.

**Comment by gatecat** [comment]
> Unfortunately, the fpga is unable to run successfully the bitstream to sram and revert to previously loaded bitstream in flash produced by quartus.

Can you provide a bit of info on how you are generating and loading the bitstream?

The CMUX stuff can't be directly copied across because the two tools might have routed the clock differently, this depends on other placement and routing decisions. Whilst I can't rule out a problem there, I don't expect it would cause the bitstream to be totally rejected either.

**Comment by infphyny** [comment]
In my makefile I have this
```
$(PROJ).rbf : $(PROJ).json Makefile
	nextpnr-mistral  -r --device $(DEVICE)  --json $< --rbf $@    --freq $(TARGET_FREQUENCY) --qsf $(QSF)


$(PROJ).json : $(FILES)
	yosys -p "read_verilog $(FILES) ; synth_intel_alm  -family cyclonev  -top $(TOP)  ; write_json $(PROJ).json"
```

I use openFPGALoader to load bitstream for both quartus and mistral.

```
openFPGALoader -m --bitstream ${PROJ}.rbf
```

**Comment by gatecat** [comment]
that all looks reasonable, I think I'll have to ask @galibert if they can see anything wrong here. Possibly the missing `IDCODE` and `JTAG_ID` writes?

**Comment by infphyny** [comment]
@gatecat you are right . I copied  IDCODE and JTAG ID from quartus  in .bt file of mistral. Then compiled .bt  to .rbf  and it work.
I will test other example to confirm.
So this is not related to CMUXVG.

**Comment by infphyny** [comment]
Very good, it is only IDCODE and JTAG missing in rbf file. The second example work with adding those 2 lines in .bt file
```
o IDCODE 22
o JTAG_ID 0078fd4b
```

**Comment by Ravenslofty** [comment]
Definitely one for @Sarayan, but should be easily fixable.

**Comment by infphyny** [comment]
To be able to program to flash, I have to put JTAG_ID 2b220dd, the  jtag idcode detected by openFPGALoader for target fpga.  When programming to sram, only "o IDCODE 22" is important.

**Comment by galibert** [comment]
I guess I have an issue somewhere with the default IDCODE.  E/A5 should
know it's 0x22...

  OG.

**Comment by galibert** [comment]
Fixed 9b0dcbe29697bb803cef4d4f6745c0e3e017e4bb

**Comment by infphyny** [comment]
Perfect, thanks.
